coprocessor/fhevm-engine/fhevm-listener/src/database/tfhe_event_propagate.rs

```diff
@@ -165,75 +193,122 @@ impl Database {
             dependencies,
             fhe_operation,
             scalar_byte,
+            log,
+            &bucket,
         )
         .await
     }
 
+    #[allow(clippy::too_many_arguments)]
     async fn insert_computation_inner(
-        &mut self,
+        &self,
         tenant_id: TenantId,
         result: &Handle,
         dependencies: Vec<Vec<u8>>,
         fhe_operation: FheOperation,
         scalar_byte: &FixedBytes<1>,
+        log: &alloy::rpc::types::Log<TfheContractEvents>,
+        bucket: &PrimitiveDateTime,
     ) -> Result<(), SqlxError> {
         let is_scalar = !scalar_byte.is_zero();
         let output_handle = result.to_vec();
-        let query = || {
-            sqlx::query!(
-                r#"
+        let query = sqlx::query!(
+            r#"
             INSERT INTO computations (
                 tenant_id,
                 output_handle,
                 dependencies,
                 fhe_operation,
-                is_scalar
+                is_scalar,
+                schedule_order,
+                transaction_id
             )
-            VALUES ($1, $2, $3, $4, $5)
+            VALUES ($1, $2, $3, $4, $5, $6, $7)
             ON CONFLICT (tenant_id, output_handle) DO NOTHING
             "#,
-                tenant_id as i32,
-                output_handle,
-                &dependencies,
-                fhe_operation as i16,
-                is_scalar
-            )
-        };
-        // retry mecanism
-        loop {
-            match query().execute(&self.pool).await {
-                Ok(_) => return Ok(()),
-                Err(err) if retry_on_sqlx_error(&err) => {
-                    error!(
-                        error = %err,
-                        "Database I/O error, will retry indefinitely"
-                    );
-                    self.reconnect().await;
-                }
-                Err(sqlx_err) => {
-                    return Err(sqlx_err);
+            tenant_id as i32,
+            output_handle,
+            &dependencies,
+            fhe_operation as i16,
+            is_scalar,
+            bucket,
+            if let Some(txh) = log.transaction_hash {
+                Some(txh.to_vec())
+            } else {
+                None
+            }
+        );
+        query.execute(&self.pool).await.map(|_| ())
+    }
+
+    async fn sort_computation_into_bucket(
+        &self,
+        output: &Handle,
+        dependencies: &[&Handle],
+        transaction_hash: &Option<Handle>,
+    ) -> PrimitiveDateTime {
+        let mut bucket_cache = self.bucket_cache.write().await;
```

Comment:
Just to think about it. Could lead to contention when this will be made concurrent. But I think It's too early to make it future proof (could do read + write lock for optimistic path when the bucket already exist, and redo a full write lock for a recheck/write or create)